lib/enterprise-event-stream.js

```diff
@@ -0,0 +1,234 @@
+ */
+EnterpriseEventStream.prototype.setStreamState = function(state) {
+	// We need to set both streamPosition and startDate, since streamPosition will be null until
+	// the first set of events is returned from Box.
+	this._streamPosition = state.streamPosition;
+	this._options.startDate = state.startDate;
+	this._options.endDate = state.endDate;
+	this._options.eventTypeFilter = state.eventTypeFilter;
+};
+
+/**
+ * Fetch the next chunk of events
+ *
+ * If there are no events, poll until events are available.
+ * If an error occurs, emit the error but continuing polling as usual.
+ * @param {Function} callback - Passed the array of events
+ * @returns {void}
+ * @private
+ */
+EnterpriseEventStream.prototype.fetchEvents = function(callback) {
+	const self = this,
+		params = {
+			stream_type: 'admin_logs'
+		};
+
+	// Use the current stream position.
+	// Handle the case where the caller passes streamPosition === 0 instead of streamPosition === '0'.
+	if (this._streamPosition || this._streamPosition === 0) {
+		params.stream_position = this._streamPosition;
+	}
+
+	if (this._options.startDate) {
+		params.created_after = this._options.startDate;
+	}
+
+	if (this._options.endDate) {
+		params.created_before = this._options.endDate;
+	}
+
+	if (this._options.eventTypeFilter) {
+		params.event_type = this._options.eventTypeFilter.join(',');
+	}
+
+	if (this._options.chunkSize) {
+		params.limit = this._options.chunkSize;
+	}
+
+	this._client.events.get(params, function(err, result) {
+		if (err) {
+			self.emit('error', err);
+			// If there was a "permanent" error, we would call the callback with it here.
+			// But it's not clear which errors are truly permanent?
+			// If Box is down or returning errors for an extended period, we still want to resume when it recovers.
+			// So, continue polling at the regular frequency.
+			// Don't use a shorter retry interval (to avoid DDOSing Box).
+		}
+
+		if (err || !result || !result.entries || result.entries.length === 0) {
+			if (!self._options.pollingInterval) {
+				// If polling is disabled, end the stream.
+				callback();
+				return;
+			}
+
+			// There were no events returned (or an error occurred), so schedule another poll.
+			const delay = self._options.pollingInterval * 1000;
+
+			// Stream readers can use this to flush buffered events to a downstream system.
+			self.emit('wait', delay);
+
+			setTimeout(function() {
+				self.fetchEvents(callback);
+			}, delay);
+			return;
+		}
+
+		// Only update the stream position if there were events returned.
+		// The API currently returns next_stream_position = 0 if there are no events (may be a bug?).
+		// But we don't want to start over at the beginning in that case, so ignore it.
+		self._streamPosition = result.next_stream_position;
+
+		// Notify the reader of the new stream position.
+		// Stream readers can respond to the 'streamPosition' event to persist the stream position.
+		self.emit('streamPosition', self._streamPosition);
```

Comment:
It seems like it would be more useful to have a `'newStreamState'` event that send the full stream state object for persistence